content/client-lib-development-guide/features.textile

```diff
@@ -1464,6 +1486,110 @@ enum StatsIntervalGranularity:
   DAY
   MONTH
 
+class DeviceDetails:
+  + getDeviceInfo() => io DeviceDetails
+  id: String
+  platform: DeviceType
+  type: DeviceType
+  clientId: String?
+  metadata: Dict<String, String>
+  updateTokenId: String
+  resetId(String)
+  resetUpdateToken() => io
+  push: DevicePushDetails
+
+class DevicePushDetails:
+  transportType: DevicePushTransportType
+  transportId: String
+  transportToken: String?
+  state: .Active | .Failing | .Failed
+  errorReason: ErrorInfo?
+  metadata: Dict<String, String>
+
+class Push:
+  publish(clientId: String, JsonObject) => io
+  publish(deviceId: String, JsonObject) => io
+  publish(transportType: String, transportId: String, transportTokenOrKey?: String) => io
+  admin: PushAdmin
+
+  // Only on platforms that support receiving notifications:
+
+  activate(DeviceDetails) => io ErrorInfo? // RSH2c
+  activateExternally(
+    DeviceDetails,
+    registerDevice: (ErrorInfo?, DeviceDetails?) -> io String?,
+    renewRegistration: (ErrorInfo?, DeviceDetails?) -> io String?
+  )
+  deactivate(deviceId: String) => io ErrorInfo?
+  deactivateExternally
+    deviceId: String,
+    deactivate: (ErrorInfo?, deviceId: String) -> io
+  )
+
+class PushAdmin:
+  deviceRegistrations: PushDeviceRegistrations
+  channelSubscriptions: PushChannelSubscriptions
+
+class JsonObject:
+  // Platform-dependant, typically a Dict-like object
+
+class PushDeviceRegistrations:
+  save(PushDeviceRegistration) => io
+  find(type: PushDeviceType, transportId: String) => io PushDeviceRegistration?
+  get(transportIds: [(PushDeviceType, String)], state?: .Active | .Failed) -> io PaginatedResult<PushDeviceRegistration>device
+  get(clientId: String, state?: .Active | .Failed) -> io PaginatedResult<PushDeviceRegistration>
+  remove(type: PushDeviceType, transportId: String) -> io
+  remove(clientId: String) => io
+  remove(PushDeviceRegistration) => io
+
+class PushChannelSubscriptions:
+  save([PushChannelSubscription]) => io
+  save(PushChannelSubscription) => io
+  listChannels() => io PaginatedResult<String>
+  get(channelName: String) => io PaginatedResult<PushChannelSubscription>
+  get(type: PushDeviceType, transportId: String, channelName: String?) => io PaginatedResult<PushChannelSubscription>
+  get(clientId: String, channelName: String?) => io PaginatedResult<PushChannelSubscription>
+  remove([PushChannelSubscription]) => io
+  remove(PushChannelSubscription) => io
+
+class PushDeviceRegistration:
```

Comment:
I am not following why this exists.  PushDeviceRegistration, as far as I can tell, is just `DeviceDetails` with the `push` attributes flattened.  We don't have `ApiKey` and `ApiKeyRegistration` simply because it's persisted to disk.  As far as I am concerned, it's just `DeviceDetails` which may be an object in memory that is not yet persisted, or an object in the Ably system that is persisted.  Added to which, should that then not just be `Device`?